pandas/core/indexes/base.py

```diff
@@ -3244,9 +3244,9 @@ def join(self, other, how='left', level=None, return_indexers=False,
         elif how == 'right':
             join_index = other
         elif how == 'inner':
-            join_index = self.intersection(other, sort=sort)
+            join_index = self.intersection(other, sort=False)
         elif how == 'outer':
-            join_index = self.union(other, sort=sort)
+            join_index = self.union(other)
```

Comment:
On Master ``union`` does generally sort by default so I have left the default ``sort=True`` here for now for compatibility reasons so the current behaviour of ``join(..., how='outer')`` does not change for now and all of the tests expecting the results of this type of join to be sorted don't break.

Reply:
add a TODO